File: UDF Dönüştürücü/Core/Tools/MergeService.swift
import Foundation
import PDFKit

enum MergeError: LocalizedError {
    case cannotOpenFile(String)
    case noPages
    case writeFailed

    var errorDescription: String? {
        switch self {
        case .cannotOpenFile(let name):
            return "\(name) dosyası açılamadı."
        case .noPages:
            return "Birleştirilecek sayfa bulunamadı."
        case .writeFailed:
            return "Birleştirilmiş PDF kaydedilemedi."
        }
    }
}

/// Birden fazla UDF/PDF dosyasını tek bir PDF'te birleştirir.
/// UDF girdileri önce mevcut dönüştürme motoruyla PDF'e çevrilir.
final class MergeService {

    /// Verilen sıradaki dosyaları tek PDF'te birleştirir ve çıktı URL'sini döner.
    static func merge(urls: [URL], progress: ((Int, Int) -> Void)? = nil) throws -> URL {
        let merged = PDFDocument()

        for (index, url) in urls.enumerated() {
            progress?(index, urls.count)

            let pdfURL: URL
            if url.pathExtension.lowercased() == "udf" {
                let document = try UDFParser.parse(fileURL: url)
                pdfURL = try PDFConverter.convert(document: document)
            } else {
                pdfURL = url
            }

            guard let pdf = PDFDocument(url: pdfURL) else {
                throw MergeError.cannotOpenFile(url.lastPathComponent)
            }
            for pageIndex in 0..<pdf.pageCount {
                if let page = pdf.page(at: pageIndex) {
                    merged.insert(page, at: merged.pageCount)
                }
            }
        }

        guard merged.pageCount > 0 else { throw MergeError.noPages }

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmm"
        let outputURL = PDFConverter.outputDirectory()
            .appendingPathComponent("Birlestirilmis_\(formatter.string(from: Date()))")
            .appendingPathExtension("pdf")

        guard merged.write(to: outputURL) else { throw MergeError.writeFailed }
        progress?(urls.count, urls.count)
        return outputURL
    }
}

File: app/src/main/java/com/velikececi/udfdonusturucu/core/tools/MergeService.kt
package com.velikececi.udfdonusturucu.core.tools

import android.content.Context
import com.tom_roush.pdfbox.io.MemoryUsageSetting
import com.tom_roush.pdfbox.multipdf.PDFMergerUtility
import com.velikececi.udfdonusturucu.core.converters.OutputPaths
import com.velikececi.udfdonusturucu.core.converters.PdfBoxInit
import com.velikececi.udfdonusturucu.core.converters.PdfConverter
import com.velikececi.udfdonusturucu.core.parser.UdfParser
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

sealed class MergeException(message: String) : Exception(message) {
    class CannotOpenFile(name: String) : MergeException("\"$name\" dosyası açılamadı.")
    class NoPages : MergeException("Birleştirilecek sayfa bulunamadı.")
    class WriteFailed(detail: String?) : MergeException("Birleştirme başarısız: ${detail ?: "bilinmeyen hata"}")
}

/**
 * MergeService.swift'in Kotlin karşılığı. UDF girdiler önce [UdfParser] + [PdfConverter] ile
 * PDF'e çevrilir, ardından tüm PDF'ler `PDFMergerUtility` ile tek dosyada birleştirilir.
 *
 * iOS tarafı sayfa sayfa `PDFDocument.insert(page, at:)` döngüsü kullanabiliyordu (PDFKit COS
 * nesnelerini kendi kopyalıyor); PDFBox'ta bunun karşılığı YOKTUR — `PDDocument.addPage(kaynak
 * sayfa)` yalnızca sayfa sözlüğünü kopyalar, dolaylı referanslar kaynak belgenin COS havuzuna
 * bakmaya devam eder ve çıktı bozuk olur. Bu yüzden kaynak nesneleri doğru klonlayan
 * `PDFMergerUtility.appendDocument` (üstündeki `mergeDocuments`) kullanılır.
 */
object MergeService {

    /**
     * [files] en az 2 eleman içermeli. [onProgress] her girdi işlenmeden hemen önce
     * `(sıraNo, toplam)` ile çağrılır — UDF→PDF ön dönüşümü dahil ilerlemeyi yansıtır.
     */
    fun merge(files: List<File>, context: Context, onProgress: ((current: Int, total: Int) -> Unit)? = null): File {
        require(files.size >= 2) { "En az 2 dosya gerekli." }
        PdfBoxInit.ensure(context)

        val pdfInputs = files.mapIndexed { index, file ->
            onProgress?.invoke(index + 1, files.size)
            when (file.extension.lowercase(Locale.ROOT)) {
                "udf" -> try {
                    val document = UdfParser.parse(file)
                    PdfConverter.convert(document, context)
                } catch (e: Exception) {
                    throw MergeException.CannotOpenFile(file.name)
                }
                else -> file
            }
        }

        val outputFile = File(
            OutputPaths.outputDirectory(context),
            "Birlestirilmis_${SimpleDateFormat("yyyyMMdd_HHmm", Locale("tr")).format(Date())}.pdf",
        )

        try {
            val merger = PDFMergerUtility().apply {
                destinationFileName = outputFile.absolutePath
                acroFormMergeMode = PDFMergerUtility.AcroFormMergeMode.PDFBOX_LEGACY_MODE
                isIgnoreAcroFormErrors = true
            }
            pdfInputs.forEach { merger.addSource(it) }
            merger.mergeDocuments(MemoryUsageSetting.setupTempFileOnly().setTempDir(context.cacheDir))
        } catch (e: MergeException) {
            throw e
        } catch (e: Exception) {
            throw MergeException.WriteFailed(e.message)
        }

        if (!outputFile.exists() || outputFile.length() == 0L) {
            throw MergeException.NoPages()
        }

        return outputFile
    }
}
